src/embit/bip21.py

```diff
@@ -0,0 +1,276 @@
+        """Parse a single parameter name/value pair"""
+        if name == self.FIELD_AMOUNT and value:
+            try:
+                # Parse amount as decimal with up to 8 decimal places
+                amount_decimal = Decimal(value.replace(',', '.'))
+                if amount_decimal < 0:
+                    raise ValueError("Negative amount")
+                
+                # Convert to satoshis
+                satoshis = int(amount_decimal * self.SATOSHIS_PER_BITCOIN)
+                if satoshis > self.MAX_BITCOIN * self.SATOSHIS_PER_BITCOIN:
+                    raise ValueError("Amount exceeds maximum")
+                
+                self._put_with_validation(self.FIELD_AMOUNT, satoshis)
+            except (InvalidOperation, ValueError, OverflowError) as e:
+                raise BIP21Error(f"'{value}' is not a valid amount: {e}")
+        
+        elif name == self.FIELD_BC and value:
+            # Parse bc parameter (Bech32m address) and add to list
+            try:
+                decoded_bc = unquote(value)
+                # Basic validation for Bech32m address format (case-insensitive)
+                if not decoded_bc.lower().startswith('bc1'):
+                    raise ValueError("Bech32m address must start with 'bc1'")
+                # Add to list of bc addresses (normalized to lowercase)
+                self.parameters[self.FIELD_BC].append(decoded_bc.lower())
+            except Exception as e:
+                raise BIP21Error(f"'{value}' is not a valid Bech32m address: {e}")
+        
+        elif name == self.FIELD_SILENT_PAYMENT and value:
+            # Parse silent payment address and add to list
+            try:
+                decoded_sp = unquote(value)
+                # Basic validation for silent payment address format
+                if not decoded_sp.startswith('sp1'):
+                    raise ValueError("Silent payment address must start with 'sp1'")
+                # Add to list of silent payment addresses
+                self.parameters[self.FIELD_SILENT_PAYMENT].append(decoded_sp.lower())
+            except Exception as e:
+                raise BIP21Error(f"'{value}' is not a valid silent payment address: {e}")
+        
+        elif name.startswith("req-"):
+            # Required parameter we don't know about
+            raise BIP21Error(f"'{name}' is required but not known, this URI is not valid")
+        
+        else:
+            # Known or unknown optional parameter
+            if value:
+                decoded_value = unquote(value)
+                self._put_with_validation(name, decoded_value)
+    
+    def _put_with_validation(self, key, value):
+        """Add parameter with duplicate validation for non-payment instruction fields"""
+        if key in self.parameters and key not in self.PAYMENT_INSTRUCTION_FIELDS:
+            raise BIP21Error(f"'{key}' is duplicated, URI is invalid")
+        self.parameters[key] = value
+    
+    def get_address(self):
```

Comment:
add types. something like this, from here onward
```suggestion
    def get_address(self) -> str | None:
```

Reply:
It's a good suggestion. A possible TODO for the future is to improve code consistency by adding type hints, as I think `embit`'s code is currently quite messy in this regard.